File: core/instruction.c
#include "instruction.h"
#include "executable.h"

int instructionGrp(Instruction* instruction, x86_insn_group grp)
{
    size_t n;
    for(n = 0;n < instruction->instruction.detail->groups_count;n++)
        if(instruction->instruction.detail->groups[n] == grp)
            return 1;

    return 0;
}

void instructionSetup(Instruction* instruction)
{
    instruction->numberOperand = instruction->instruction.detail->x86.op_count;
    instruction->operands = malloc(sizeof(Instruction)*instruction->instruction.detail->x86.op_count);
    size_t i;
    for(i = 0;i < instruction->instruction.detail->x86.op_count;i++)
    {
        Operand op;
        op.op = instruction->instruction.detail->x86.operands[i];
        op.instruction = instruction;
        op.label = 0;
        if(instructionGrp(instruction, X86_GRP_JUMP))
        {
            op.label = malloc(10);
            sprintf(op.label, ".L%x", op.op.imm);
        }
        else if(instructionGrp(instruction, X86_GRP_CALL))
        {
            size_t j;
            for(j = 0;j < instruction->function->exe->numberFunction;j++)
                if(instruction->function->exe->functions[j].address == op.op.imm)
                {
                    op.label = instruction->function->exe->functions[j].name;
                }
        }

        instruction->operands[i] = op;
    }
}
char* instructionToString(Instruction* instruction)
{
    char* result = malloc(100);
    strcpy(result, "");
    if(instruction->label != 0)
    {
        strcat(result, instruction->label);
        strcat(result, ":\n");
    }
    strcat(result, "\t");
    strcat(result, instruction->instruction.mnemonic);
    strcat(result, " ");

    size_t i;
    for(i = 0;i < instruction->numberOperand;i++)
    {
        strcat(result, operandToString(&instruction->operands[i]));
        strcat(result, ", ");
    }

    result[strlen(result)-2] = 10;
    result[strlen(result)-1] = 0;
    result[strlen(result)] = 0;
    return result;
}

File: IOT/core/instruction.py
class Instruction():
    CALL = 0
    JUMP = 1
    def __init__(self, insn, operands, function):
        self.function = function
        self.insn = insn
        self.address = insn.address
        self.size  = insn.size
        self.type = -1
        self.mnemonic = insn.mnemonic
        self.operands = operands
        for op in self.operands:
            op.instruction = self
        self.label = ".L" + hex(self.address)

    def setup(self):
        for operand in self.operands:
            if self.mnemonic[0] == 'j' and self.operands[0].label == "":# or self.mnemonic[:4] == "call":
                self.operands[0].label = ".L" + hex(self.operands[0].op.imm)
            else:
                if self.mnemonic[:4] == "call":
                    for f in self.function.executable.functions:
                        if f.address == self.operands[0].op.imm:
                            self.operands[0].label = f.name

    def __str__(self):
        str = ""
        if self.label != "":
            str = self.label + ":"
        str = str + "\t"+self.mnemonic+"\t\t"
        for operand in self.operands:
            str = str + operand.__str__() + ", "

        str = str[0:len(str)-2] + "\n"
        return str